clippy_lints/src/useless_allocation.rs

```diff
@@ -0,0 +1,119 @@
+use rustc_errors::Applicability;
+use rustc_hir::{BorrowKind, Expr, ExprKind, LangItem};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{self, ClauseKind, ImplPolarity, Mutability, Ty};
+use rustc_session::declare_lint_pass;
+use rustc_span::sym;
+
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::is_diag_trait_item;
+use clippy_utils::source::snippet_opt;
+use clippy_utils::ty::{is_type_diagnostic_item, is_type_lang_item};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks if an unneeded allocation is performed when trying to get information
+    /// related to a given key is a `HashMap`-like type.
+    ///
+    /// ### Why is this bad?
+    /// Using less resources is generally a good idea.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let mut s = HashSet::from(["a".to_string()]);
+    /// s.remove(&"b".to_owned());
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let mut s = HashSet::from(["a".to_string()]);
+    /// s.remove("b");
+    /// ```
+    #[clippy::version = "1.78.0"]
+    pub USELESS_ALLOCATION,
```

Comment:
issue: I think this lint name is quite general for how specific the actual lint is

Reply:
its a bit as if `needless_clone` and `unnecessary_to_owned` had a child together.

Reply:
That makes me wonder if we don't just want to extend `unnecessary_to_owned` with these additional cases.

Reply:
For example, `unnecessary_to_owned` already lints on:
```rs
fn f(x: &str) {}
f(&"".to_owned());
```

Reply:
> its a bit as if `needless_clone` and `unnecessary_to_owned` had a child together.

Pretty much yes. :laughing: 

@y21 I wondered about that as well but since it's covering than just `to_owned` call, I felt like it should not be there. Same goes for `needless_clone` since it's not related to `clone`.

Reply:
So what do you want me to do here? Rename it or merge it with another existing lint?

Reply:
In my opinion, merging it with `unnecessary_to_owned` makes the most sense.

> I wondered about that as well but since it's covering than just to_owned call

But the lint description for it also says "[...] and other to_owned-like functions", and also covers the same `to_owned`, `to_vec` and `to_string` functions.

Reply:
I also tested to add the `Borrow` trait as check and it doesn't detect this code:

```rust
fn borrow_str<Q: Borrow<str>>(s: &Q) {}
fn borrow_slice<Q: Borrow<[u8]>>(s: &Q) {}
fn check_borrow() {
    borrow_str(&"a".to_owned());
    borrow_str(&"a".to_string());
    borrow_slice(&[0].to_vec());
}
```

So I'll merge this code with the lint for now.